javaslang/src/main/java/javaslang/collection/BitMappedTrie.java

```diff
@@ -131,11 +134,33 @@ private static int treeSize(int branchCount, int depthShift) {
         } else if (n <= 0) {
             return empty();
         } else {
-            final Object[] root = modifyLeaf(array, depthShift, (offset + n) - 1, Arrays::copyTake, Arrays::copyTake);
-            return new BitMappedTrie<>(root, offset, n, depthShift);
+            final int index = n - 1;
+            final Object[] root = arePointingToSameLeaf(index, length() - 1)
+                                  ? array
+                                  : modifyLeaf(array, depthShift, offset + index, Arrays::copyTake, IDENTITY);
+            return collapsed(root, offset, n, depthShift);
         }
     }
 
+    private boolean arePointingToSameLeaf(int i, int j) {
+        final boolean result = firstDigit(offset + i, BRANCHING_BASE) == firstDigit(offset + j, BRANCHING_BASE);
+        assert result == (getLeaf(i) == getLeaf(j));
+        return result;
+    }
+
+    /* drop root node while it has a single element */
+    private static <T> BitMappedTrie<T> collapsed(Object[] array, int offset, int length, int shift) {
```

Comment:
Please explain to me how nulls can occur. I have questions:

* Does array contain these nulls?
* What is the meaning of these nulls?
* Is more and more memory consumed over time be nulled array slots?

The BMT should only contain valid nodes.

Before changing the implementation please explain to me how the tree structure changes over time.